src/example/SampleCustomLoader/sampleCustLoader.cpp

```diff
@@ -0,0 +1,436 @@
+        watcher_thread = std::move(th);
+        watcherStarted = true;
+    }
+    std::cout << "custSampleLoader: StartWatcher" << std::endl;
+}
+
+void custSampleLoader::watcherJoin() {
+    std::cout << "custSampleLoader: watcherJoin()" << std::endl;
+    if (watcherStarted) {
+        if (watcher_thread.joinable()) {
+            watcherStarted = false;
+            cv.notify_all();
+            watcher_thread.join();
+        }
+    }
+}
+
+/* 
+ * From the custom loader options extract the mdeol file name and other neede information and
+ * load the model and optional bin file into buffers and return
+ */
+CustomLoaderStatus custSampleLoader::loadModel(const std::string& modelName, const std::string& basePath, const int version,
+    const std::string& loaderOptions, char** xmlBuffer, int* xmlLen,
+    char** binBuffer, int* binLen) {
+    std::cout << "custSampleLoader: Custom loadModel" << std::endl;
+
+    std::string binFile;
+    std::string modelFile;
+    std::string enableFile;
+    int modelType = SAMPLE_LOADER_IR_MODEL;
+    CustomLoaderStatus st = CustomLoaderStatus::MODEL_LOAD_ERROR;
+
+    int ret =
+        extract_input_params(basePath, version, loaderOptions, binFile, modelFile, enableFile, modelType);
+    if (ret != SAMPLE_LOADER_OK) {
+        std::cout << "custSampleLoader: Invalid custom loader options" << std::endl;
+        return st;
+    }
+
+    // load models
+    ret = load_files(binFile, modelFile, modelType, xmlBuffer, binBuffer, xmlLen, binLen);
+    if (ret != SAMPLE_LOADER_OK || *xmlBuffer == NULL) {
+        std::cout << "custSampleLoader: Could not read model files" << std::endl;
+        return CustomLoaderStatus::INTERNAL_ERROR;
+    }
+
+    /* Start the watcher thread after first moel load */
+    if (watcherStarted == false) {
+        startWatcher(MODEL_CHECK_PERIOD);
+        std::this_thread::sleep_for(std::chrono::milliseconds(100));
+    }
+
+    auto modelId = std::make_pair(modelName, version);
+    models_loaded.emplace_back(modelId);
+
+    if (std::find(models_watched.begin(), models_watched.end(), modelId) == models_watched.end()) {
+	    models_watched.emplace_back(modelId);
+    }
+
+    if (!(enableFile.empty())) {
+        models_enable.insert({std::make_pair(modelName, version), enableFile});
+    }
+    if (modelType == SAMPLE_LOADER_IR_MODEL)
+        st = CustomLoaderStatus::MODEL_TYPE_IR;
+    else if (modelType == SAMPLE_LOADER_ONNX_MODEL)
+        st = CustomLoaderStatus::MODEL_TYPE_ONNX;
+    else if (modelType == SAMPLE_LOADER_BLOB_MODEL)
+        st = CustomLoaderStatus::MODEL_TYPE_BLOB;
+
+    return st;
+}
+
+// Retire the model
+CustomLoaderStatus custSampleLoader::retireModel(const std::string& modelName) {
+    std::vector<model_id_t> newVec;
+
+    for (auto it = std::begin(models_watched); it != std::end(models_watched); it++) {
+	    if (it->first != modelName) {
+		    newVec.emplace_back(*it);

```

Comment:
Why create new vector when you could just erase in place?

Reply:
The original vector may be accesses by the checkModelStatus thread. To limit the time mutex is held, add newVec